### ISSUE TITLE
Reading scenarios from log files

### ISSUE THREAD

**Issue by allegroCoder** [tuura/scenco#23]
`SCENCO` must be able to read Scenarios in the form of event log:

Example from the file `telecom.cpog`:
```
g1 = e0 -> e1 + e1 -> e10 + e10 -> e7 + e7 -> e8 + e8 -> e3
g2 = e0 -> e1 + e1 -> e11 + e11 -> e12 + e12 -> e13 + e13 -> e3
g3 = e0 -> e1 + e1 -> e12 + e10 -> e3 + e12 -> e10
g4 = e0 -> e1 + e1 -> e12 + e11 -> e3 + e12 -> e13 + e13 -> e11
g5 = e0 -> e1 + e1 -> e15 + e12 -> e24 + e15 -> e12
```

**Comment by snowleopard** [comment]
We should probably do this in two steps:
* Implement parsing of unconditional expressions.
* Add support for conditions, e.g. `g1 = a -> ([x]b + [!x]c)`. This will require parsing Boolean functions, which is more complicated.

@jrbeaumont implemented parsing for his algebra plugin in Workcraft, so perhaps you could work together on this.

Also, this is probably blocked by #4, as expressions need to be parsed into `Graph` datatype.

**Comment by snowleopard** [comment]
An alternative solution would be to use @mvdan's `hint` to simply interpret `.cpog` files as Haskell programs constructing graphs. That could be a lot easier to implement and will provide support for powerful feautres like using functions when defining graphs, e.g.:
```haskell
ifthenelse x t f = [x]t + [!x]e

g1 = a ~> ifthenelse x b c
```

**Comment by allegroCoder** [comment]
What I have done so far has been using the old scenario formalism as the final conversion end-point. It supports conditions either. Hence, as we discussed before, I would keep that without modifying the way `SCENCO (c++ side)`  read scenarios.
```
.scenario CALU_Rn_to_Rn
PCIU IFU
ALU ALU2
:!ge ALU2
.end
```
First, I want to make sure that the big event logs can still be processed in the new version of `SCENCO`, it's not been tested yet and I am about to do this.

--------

Tomorrow, I will have a word with @jrbeaumont  and @mvdan about this. To recap: I agree with you in implementing the following:

1 - Reading unconditional / conditional expression through the way you mentioned.
2 - Building graphs from them.
3 - Introducing scenarios into `SCENCO (c++ side)` with the old formalism.

**Comment by snowleopard** [comment]
> What I have done so far has been using the old scenario formalism as the final conversion end-point. It supports conditions either. 

Although it is nice to support old `.scenario` syntax, I don't think we should necessarily do that. We do not have that many files written in it to be worried about backward compatibility. So, I would be in favour of eventually supporting only the algebraic notation.

**Comment by allegroCoder** [comment]
That's true, indeed I did not mean to support the old `.syntax` formalism externally. I would use that just for transferring the graphs into the `SCENCO (c++ side)`. The user will not know about that.

Externally, `SCENCO (Haskell side)` will be able to read graphs in the form of expressions(`g1 = a -> ([x]b + [!x]c)`) as you mentioned in your previous comment.

**Comment by snowleopard** [comment]
I see. This makes sense as a first step.

**Comment by allegroCoder** [comment]
I had a chat with @mvdan and @jrbeaumont . I am still not sure which solution I am going for, however I will first think about a Graph representation that we can use for representing partial orders (graphs!).

To do that, I will probably ispire by Data Type lectures by Wadler. In some of his lecture he shows how to represent trees, I might be able to develop something similar and maybe upon that I will come up with my own method to define partial orders.

I will keep you updated, sorry in advance if it is going to take me a lot of time, unfortunately I am still very slow at Haskell coding. I will do my best!
